### ISSUE TITLE
New Source: Paypal Reports

### ISSUE THREAD

**Issue by sherifnada** [airbytehq/airbyte#2801]
## Tell us about the new integration you’d like to have
A user (jtalmi ) has indicated interest in pulling data from the [Paypal Reports API](https://developer.paypal.com/docs/reports/payflow-reports/use-the-api/)

## Describe the context around this new integration
<TODO from the user>

## Describe the alternative you are considering or using
Currently the only alternative is writing a custom script since other ELT providers don't offer this connector

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368158845034) by [Unito](https://www.unito.io)

**Comment by sherifnada** [comment]
@jtalmi could you comment about the following: 
1. Can you speak to the schema that would be most consumable for this report data? Rough outline is fine, doesn't have to be exact schema e.g: should each report type land in a different table? Which reports would you be most interested in pulling? 
2. What is the interface/UX you would expect for the reporting API? What would be the inputs/configurations you expect to supply? 
3. Would you want to pull these reports on a recurring, incremental basis (e.g: once a day) or would you only need to do this in very specified batches? 

Any other context you can provide is super helpful.

**Comment by karinakuz** [comment]
**Integration Vetting**


**Webhook-based? (no/partially/yes)**
if partially, mention which endpoints are webhook-based
_Partially (no events for Payflow Reports) https://developer.paypal.com/docs/api-basics/notifications/webhooks/event-names/_ 

**Available authentication modes (API key/Oauth/other)**
_OAuth 2.0_

**Connecting to the data source**
what kind of API is it? (SOAP/GraphQL/REST/other)
_Payflow Gateway has an XML API which enables you to develop web and desktop applications that create, schedule, and run standard (pre-defined) and custom Payflow services reports.
https://developer.paypal.com/docs/reports/payflow-reports/_

is there an SDK available? Can we connect with HTTP requests directly?
_To submit a request to the Reporting API, you post your XML request through HTTPS to the host URLs in the body of the request. No SDK is needed since you just perform a standard HTTPS post using the language of your choice. The Content-Type for the request should be text/plain._

If this is a database/datalake/pubsub source, does it have a JDBC driver or Java SDK?
_not for Payflow Reports_

**Creating an account**
Select one of the below options
self-service free account available?_ If yes then create it using integration-test@airbyte.io (you should be getting emails sent to this email address)
_yes, Sandbox_

**How to populate the account with data?**

Can it be done entirely through the API?
_yes_

Will it require some work through the UI? If so, what is required?
_no_

Will it require performing some paid action like running an Ad campaign? If so, please note the action needed, estimated cost, and estimated effort in hours. If a task will take more than 4 hours, please break it down into smaller tasks each of which take no more than 4 hours
_no_

Do you need any actions from Airbyte to populate the account with data? If so, please tag @ sherifnada for support
_no_

**Comment by karinakuz** [comment]
**Available streams for sync**

If this is an API connector:
1. Provide a link  to the reference docs
**Use the Reporting API** https://developer.paypal.com/docs/reports/payflow-reports/use-the-api/
**Samples** https://developer.paypal.com/docs/reports/payflow-reports/samples/
**Report Parameters** https://developer.paypal.com/docs/reports/payflow-reports/report-parameters/

**Main streams - Standard Reports:**

**_1. Activity report_**
POST https://payments-reports.paypal.com/test-reportingengine
Body
<?xml version="1.0" encoding="UTF-8"?>
<reportingEngineRequest>
   <authRequest>
      <user>Airbyte</user>
      <vendor>Airbyte</vendor>
      <partner>PayPal</partner>
      <password>PPAirbyte1</password>
   </authRequest>
   <getDataRequest>
      <reportId>RE0426507835</reportId>
      <pageNum>1</pageNum>
   </getDataRequest>
</reportingEngineRequest>

**_2. Settlement report_**
POST https://payments-reports.paypal.com/test-reportingengine
Body
<?xml version="1.0" encoding="UTF-8"?>
<reportingEngineRequest>
   <authRequest>
      <user>Airbyte</user>
      <vendor>Airbyte</vendor>
      <partner>PayPal</partner>
      <password>PPAirbyte1</password>
   </authRequest>
   <getDataRequest>
      <reportId>RE0426523801</reportId>
      <pageNum>1</pageNum>
   </getDataRequest>
</reportingEngineRequest>

**_3. Transactions Summary report_**
POST https://payments-reports.paypal.com/test-reportingengine
Body
<?xml version="1.0" encoding="UTF-8"?>
<reportingEngineRequest>
   <authRequest>
      <user>Airbyte</user>
      <vendor>Airbyte</vendor>
      <partner>PayPal</partner>
      <password>PPAirbyte1</password>
   </authRequest>
   <getDataRequest>
      <reportId>RE0426508292</reportId>
      <pageNum>1</pageNum>
   </getDataRequest>
</reportingEngineRequest>

**_4. Shipping and Billing report_**
POST https://payments-reports.paypal.com/test-reportingengine
Body
<?xml version="1.0" encoding="UTF-8"?>
<reportingEngineRequest>
   <authRequest>
      <user>Airbyte</user>
      <vendor>Airbyte</vendor>
      <partner>PayPal</partner>
      <password>PPAirbyte1</password>
   </authRequest>
   <getDataRequest>
      <reportId>RE0426523863</reportId>
      <pageNum>1</pageNum>
   </getDataRequest>
</reportingEngineRequest>

**_5. Batch ID report_**
POST https://payments-reports.paypal.com/test-reportingengine
Body
<?xml version="1.0" encoding="UTF-8"?>
<reportingEngineRequest>
   <authRequest>
      <user>Airbyte</user>
      <vendor>Airbyte</vendor>
      <partner>PayPal</partner>
      <password>PPAirbyte1</password>
   </authRequest>
   <getDataRequest>
      <reportId>RE0426524088</reportId>
      <pageNum>1</pageNum>
   </getDataRequest>
</reportingEngineRequest>

_**6. Audit Activity report**_
POST https://payments-reports.paypal.com/test-reportingengine
Body
<?xml version="1.0" encoding="UTF-8"?>
<reportingEngineRequest>
   <authRequest>
      <user>Airbyte</user>
      <vendor>Airbyte</vendor>
      <partner>PayPal</partner>
      <password>PPAirbyte1</password>
   </authRequest>
   <getDataRequest>
      <reportId>RE0426524182</reportId>
      <pageNum>1</pageNum>
   </getDataRequest>
</reportingEngineRequest>

2. Please mention if any of the endpoints are asynchronous (meaning we have to make a request, wait for the server to compute a response, then make another request to obtain the result)
 
if this is a source where the schema is dynamic/not fixed e.g: database, mention the data model of the connector i.e: relational, document-based, graph-based, etc.
**_yes_**

**Do any streams support incremental sync?**
not necessary to mention all of them, just saying “none”, “some”, or “all” is sufficient. If only some streams, please mention which ones support incremental sync
**_no_**

**Preparation scope & blockers**
What information, materials, or anything else you will need from Airbyte in order to perform preparation for this connector? 
_**no**_

**Other information/blockers**
place any other useful information you learned here
**_no_**

**Comment by karinakuz** [comment]
Populating the account with data 
[ ] Do activities on the account: open/close 10 times, add transactions (UI), create API queries in Postman to create the reports and get data of the reports. Add notes to LastPass Description with name Payflow Reports (12h).

Get correct data in responses for POST XML requests:
1. Activity report
POST https://payments-reports.paypal.com/test-reportingengine
Body
<?xml version="1.0" encoding="UTF-8"?>
<reportingEngineRequest>
   <authRequest>
      <user>Airbyte</user>
      <vendor>Airbyte</vendor>
      <partner>PayPal</partner>
      <password>PPAirbyte1</password>
   </authRequest>
   <getDataRequest>
      <reportId>RE0426507835</reportId>
      <pageNum>1</pageNum>
   </getDataRequest>
</reportingEngineRequest>

Airbyte Airbyte PayPal PPAirbyte1 RE0426507835 1
2. Settlement report
POST https://payments-reports.paypal.com/test-reportingengine
Body

Airbyte Airbyte PayPal PPAirbyte1 RE0426523801 1
3. Transactions Summary report
POST https://payments-reports.paypal.com/test-reportingengine
Body

Airbyte Airbyte PayPal PPAirbyte1 RE0426508292 1
4. Shipping and Billing report
POST https://payments-reports.paypal.com/test-reportingengine
Body

Airbyte Airbyte PayPal PPAirbyte1 RE0426523863 1
5. Batch ID report
POST https://payments-reports.paypal.com/test-reportingengine
Body

Airbyte Airbyte PayPal PPAirbyte1 RE0426524088 1
6. Audit Activity report
POST https://payments-reports.paypal.com/test-reportingengine
Body

Get correct data in responses for GET requests:
Get Events https://api2.frontapp.com/events
Get Responses https://api2.frontapp.com/responses
Get Contacts https://api2.frontapp.com/contacts
Get Channels https://api2.frontapp.com/channels
Get Inboxes https://api2.frontapp.com/inboxes
Get Comments (get data for certain conversation only) https://api2.frontapp.com/conversations/conversation_id/comments
Get Conversations https://api2.frontapp.com/conversations
Get Messages (get data for certain conversation only) https://api2.frontapp.com/conversations/conversation_id/messages
Get Rules https://api2.frontapp.com/rules
Get Tags https://api2.frontapp.com/tags
Get Teams https://api2.frontapp.com/teams
Get Teammates https://api2.frontapp.com/teammates
Get Topics https://api2.frontapp.com/topics

**Comment by natikgadzhi** [comment]
@pnilan oh look, an XML connector!

**Comment by natikgadzhi** [comment]
TLDR now that we have XML format support in Connector Builder, making this connector should be possible, and relatively easy.
